contracts/Minter.sol

```diff
@@ -0,0 +1,164 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.11;
+
+interface IMintable {
+  function mint(address to, uint256 amount) external;
+  function decimals() external returns (uint256);
+}
+
+/// @title A Minter contract for Splinterlands
+/// @author Splinterlands Team (@fbslo)
+
+contract Minter {
+  /// @notice Address that can change pools
+  address public admin;
+  /// @notice Address of the token to mint
+  address public token;
+  /// @notice Block number when mint() was last called
+  uint256 public lastMintBlock;
+  /// @notice Total number of tokens already minted
+  uint256 public totalMinted;
+  /// @notice Maximum number of tokens minted, 3B (with 18 decimal places)
+  uint256 public cap = 3000000000000000000000000000;
+  /// @notice Maximum number of pools
+  uint256 public poolsCap = 100;
+  /// @notice Maximum amount per block
+  uint256 public maxPerBlock;
+
+  /// @notice Struct to store information about each pool
+  struct Pool {
+    address receiver;
+    uint256 amountPerBlock;
+  }
+  /// @notice Array to store all pools
+  Pool[] public pools;
+
+  /// @notice Emitted when mint() is called
+  event Mint(address receiver, uint256 amount);
+  /// @notice Emitted when pool is added
+  event PoolAdded(address newReceiver, uint256 newAmount);
+  /// @notice Emitted when pool is updated
+  event PoolUpdated(uint256 index, address newReceiver, uint256 newAmount);
+  /// @notice Emitted when pool is removed
+  event PoolRemoved(uint256 index, address receiver, uint256 amount);
+  /// @notice Emitted when admin address is updated
+  event UpdateAdmin(address admin, address newAdmin);
+
+  /// @notice Modifier to allow only admin to call certain functions
+  modifier onlyAdmin(){
+    require(msg.sender == admin, '!admin');
+    _;
+  }
+
+  /**
+   * @notice Constructor of new minter contract
+   * @param newToken Address of the token to mint
+   * @param startBlock Initial lastMint block
+   * @param newAdmin Initial admin address
+   * @param newMaxPerBlock Maximum amount minted per block
+   */
+  constructor(address newToken, uint256 startBlock, address newAdmin, uint256 newMaxPerBlock){
+    require(startBlock >= block.number, "Start block must be above current block");
+    require(newToken != address(0), 'Token cannot be address 0');
+    require(newAdmin != address(0), 'Admin cannot be address 0');
+
+    token = newToken;
+    lastMintBlock = startBlock;
+    admin = newAdmin;
+    maxPerBlock = newMaxPerBlock;
+
+    emit UpdateAdmin(address(0), newAdmin);
+  }
+
+  /**
+   * @notice Mint tokens to all pools, can be called by anyone
+   */
+  function mint() public {
+    uint256 mintDifference = block.number - lastMintBlock;
+
+    for (uint256 i = 0; i < pools.length; i++){
+      uint256 amount = pools[i].amountPerBlock * mintDifference;
+
+      if(totalMinted + amount >= cap && totalMinted != cap){
+        amount = cap - totalMinted;
+      }
+      require(totalMinted + amount <= cap, "Cap reached");
```

Comment:
This code is a bit hard to reason about and I'm not sure it's doing exactly what you want. If the cap is reached in the midst of minting, then one pool will be "given" a bit less than their total amount, but on the next iteration of the loop, this require statement will be hit and the entire transaction will revert. The end result is that the effective cap ends up a bit below the total cap. The only way this would not be the case is if the `if` clause defined on line 82 is triggered on the last pool in the loop. If this is the desired behavior, it's a very confusing way to specify it.

Can you please explain what behavior is *desired* when the minted amount approaches the cap?

Reply:
Require statement should never be hit since `totalMinted + amount` can't be larger than `cap`. If the cap is hit in the middle of the loop, the remaining pools will receive 0 tokens.

I moved `require(totalMinted != cap, "Cap reached");` to the start of the function, so if the cap was reached in the previous transaction, it will revert immediately; https://github.com/steem-monsters/sps-minter/commit/877abe4f1205e3a4f288f78cc8a53d4eae04d6cd

Reply:
Looks good for the behavior described. As mentioned elsewhere, I recommend changing the comparison operator and adding tests to explicitly test when the cap is reached during the loop.